### ISSUE TITLE
Better organization of Actions in napari codebase

### ISSUE THREAD

**Issue by Czaki** [napari/napari#6516]
## 🧰 Task

The current design of `Actions` using `app-model` creations requires editing multiple files.

There is a file where the action function is defined (for example `napari/_qt/_qapp_model/qactions/_file.py`) then one needs to add an action identifier `napari/_app_model/constants/_commands.py` and a description (same file but over 100 lines below). Define Action in `napari/_qt/_qapp_model/qactions/_file.py` (but many lines below). 

Shortcuts are defined currently in Action, but it prevents them from being editable. But the rest of the shortcuts are defined in `napari/utils/shortcuts.py`

And I don't know if I found all the places that need to be edited. 

We mainly conver menus to `app-models` There are multiple actions yet handled by old ActionManager, actions from plugins (yaml defined). Actions form `Layer.bind_key()`. 

I think that the current solution is to complex. I think that we should reduce the number of places that need to be edited to two. First, the place where action is defined (function implementation). And second, when it is added to the list of registered actions (no magical add on import). 

We could adopt the old approach of using a decorator, which will add `__action__` attribute to the function. Then, based on a list of all registered actions, the list of default shortcuts will be created. 

If we have a global list of actions (that could be composited from subsist for given modules), we could check identifier collision, shortcuts collisions etc. 

So my proposition will be:

```python 
def action_dec(id: str, title: str, menus: Optional[dict], shortcut, toggled: Callback):
    def _wraps(func):
        func.__action__ = Action(id=id, title=title, menus=menus, shortcut=shortcut, toggled=toggled)
        return func
    return _wraps
```

Then put functions on list. and collect them in one place to register.

**Comment by Czaki** [comment]
During the review of #4991 I reminded that we use injection whenever trigger Action through GUI  element, even when we could bind parameters during the creation of the widget.

Also spot bug that I submit PR to fix: #6521

**Comment by lucyleeow** [comment]
There's quite a few points raised here. I will attempt to summarise and add some points. I don't have much knowledge about action manager so I don't follow some of your points.

### Actions

> I think that we should reduce the number of places that need to be edited to two

> no magical add on import

You're right, adding/changing a menu action requires changing 2 files:

1. command ID AND title in `napari/_app_model/constants/_commands.py` - requires changing this file in 2 places. I agree that it's a long file and annoying to navigate
    * I think the Enums in this file is to avoid collision so your suggestion of a global list to then check for identifier collision would avoid the need for the Enums. It could also avoid the need for the menu Enums (in `napari/_app_model/constants/_menus.py`). How would this checking for collision work? Would it be a test that fails, telling the dev that a ID/title needs to be changed?
    * I like being able to see all the IDs and titles in one place, but maybe this is not that useful
2. `Action` definition e.g., in the `napari/_qt/_qapp_model/qactions/_file.py`, sometimes the function definition is also here, sometimes it is elsewhere (e.g., `QtViewer._open_files_dialog`)

**Decorator**

Is the idea of the decorator such that Actions are 'created' when the function is defined? The Actions would then be scattered across the codebase. Atm they are at least confined to 2 locations (qt/non-qt), which I prefer.

### Keybindings/shortcuts

> Shortcuts are defined currently in Action, but it prevents them from being editable.

Good point. I do not know if Talley had a solution or plans for this :shrug: 

> But the rest of the shortcuts are defined in napari/utils/shortcuts.py

I thought the new keybinding system was meant to unify shortcuts to app model, but  @DragaDoncila says this may only be dealing with plugin contribution keybindings (summary here: #5874).

> Then, based on a list of all registered actions, the list of default shortcuts will be created.

This is possible already right?

### Dependency injection

I thought the whole purpose of in-n-out was to delay injection until runtime, as in this is specifically what we wanted? @DragaDoncila can probably explain why better.

**Comment by DragaDoncila** [comment]
Thanks for opening this issue @Czaki! Now that menu conversion is almost done I think it's a good spot to talk about the different things we want ahead of fully deprecating Action Manager.

## Actions

> There are multiple actions yet handled by old ActionManager, actions from plugins (yaml defined). Actions form Layer.bind_key().

The layer actions I was aware of, but afaik plugin actions, after merging #4991 will be entirely handled by `app-model`. Can you please expand on what plugin actions you're referring to?

> And second, when it is added to the list of registered actions (no magical add on import).

Afaik, actions are registered in two places. On `NapariApplication` [init](https://github.com/napari/napari/blob/main/napari/_app_model/_app.py#L36) and on `_QtMainWindow` [init](https://github.com/napari/napari/blob/main/napari/_qt/qt_main_window.py#L179). @lucyleeow is that correct? @Czaki where/when would you prefer this registration occur? I'm not sure what you mean by "add on import". 

> We could adopt the old approach of using a decorator, which will add __action__ attribute to the function. 

I guess I'm not particularly opposed to using a decorator, although I remember there being some discussion about how our use of decorators for binding shortcuts and callbacks made it difficult to understand how things were connected and as @lucyleeow mentions, that concern is also true of an action decorator. But again, I don't feel strongly about this.

## Keybindings

> If we have a global list of actions (that could be composited from subsist for given modules), we could check identifier collision, shortcuts collisions etc.

As far as I'm aware the `app-model` `app`'s `registries` are exactly this "global list". Identifier collision is already provided by the `command` registry. The `keybindings` registry doesn't implement collision detection, but presumably it could. `KeybindingRule`s also support a `weight`, which could allow us to register different priority keybindings for the same command ID. To me it makes more sense to use these registries, rather than have a separate registry for collision checks.

## Dependency Injection

> During the review of https://github.com/napari/napari/pull/4991 I reminded that we use injection whenever trigger Action through GUI element, even when we could bind parameters during the creation of the widget.

I'm not sure what you mean here sorry. Afaik, **when** parameters are bound hasn't really changed with `app-model`? The `viewer` is provided on widget creation in #4991 (in `toggle_or_get_widget` before [the widget's callable](https://github.com/lucyleeow/napari/blob/menu_plugin/napari/_qt/_qplugins/_qnpe2.py#L274) is called - which builds the widget), while remaining parameters usually come from the widget itself in the case of magicgui widget actions.

**Comment by Czaki** [comment]
> The layer actions I was aware of, but afaik plugin actions, after merging #4991 will be entirely handled by `app-model`. Can you please expand on what plugin actions you're referring to?

I describe here all sources of action in the current napari codebase. I wrongly phase it, so I suggest that yaml actions are not handled by app-model. But as they are not handled by the old ActionManager, we do not control shortcuts for them. 

> Afaik, actions are registered in two places. On `NapariApplication` [init](https://github.com/napari/napari/blob/main/napari/_app_model/_app.py?rgh-link-date=2023-12-19T06%3A30%3A22Z#L36) and on `_QtMainWindow` [init](https://github.com/napari/napari/blob/main/napari/_qt/qt_main_window.py?rgh-link-date=2023-12-19T06%3A30%3A22Z#L179). @lucyleeow is that correct? @Czaki where/when would you prefer this registration occur? I'm not sure what you mean by "add on import".

I describe the current behavior of the old ActionManager (still present in code), that registers action on import file using different decorators. 

We need to also allow register/unregister action on runtime, as some actions may require the presence of some widget. Or someone could want to register a custom action in a script/notebook. 


> As far as I'm aware the `app-model` `app`'s `registries` are exactly this "global list". Identifier collision is already provided by the `command` registry. The `keybindings` registry doesn't implement collision detection, but presumably it could. `KeybindingRule`s also support a `weight`, which could allow us to register different priority keybindings for the same command ID. To me it makes more sense to use these registries, rather than have a separate registry for collision checks.

This section was about removing the command registry. 

> I'm not sure what you mean here sorry. Afaik, **when** parameters are bound hasn't really changed with `app-model`? The `viewer` is provided on widget creation in #4991 (in `toggle_or_get_widget` before [the widget's callable](https://github.com/lucyleeow/napari/blob/menu_plugin/napari/_qt/_qplugins/_qnpe2.py?rgh-link-date=2023-12-19T06%3A30%3A22Z#L274) is called - which builds the widget), while remaining parameters usually come from the widget itself in the case of magicgui widget actions.

Currently, the menu are built using `QModelMenu` class from app-model. It uses injection every time use click, even if it could be resolved on constructor call. So bugs like #6466 create much more problem for end user.

**Comment by lucyleeow** [comment]
> But as they are not handled by the old ActionManager, we do not control shortcuts for them

To be clear, you mean that the user cannot change the shortcuts for app-model `Actions` (we/napari can only set one default)? 

> This section was about removing the command registry.

You can't really remove the command registry from app-model. App-model at it's heart is just 3 registries (menu, command and keybinding registries)

Talley has added some new docs to app-model which introduces app-model well, see: https://app-model--142.org.readthedocs.build/en/142/getting_started/#registries

> We need to also allow register/unregister action on runtime, as some actions may require the presence of some widget

There is an `enablement` field which lets you specify condition(s) for when the `Action` is active. However, maybe there is another use case example that is not satisfied by `Action`s?

> someone could want to register a custom action in a script/notebook.

More details needed. An action that gets added to the menu (I don't think this is possible now even)? Would this be akin to NAP-6, but you want the user to be able to do this via script (on top of via a plugin)

> It uses injection every time use click, even if it could be resolved on constructor call.

To clarify, any time user 'clicks' anywhere, or clicks on the menu item? Either way, yes I can see how this is less performant.

>  I remember there being some discussion about how our use of decorators for binding shortcuts and callbacks made it difficult to understand how things were connected

This does resonate with me. I am -0 for the decorator. As I mentioned, functions are defined in many places in the code base, I think it would be better for Actions to be in one/two places.

Potentially this could be discussed at a community meeting, I think it can make clearing things up easier.

**Comment by Czaki** [comment]
> To be clear, you mean that the user cannot change the shortcuts for app-model `Actions` (we/napari can only set one default)?

yes. And the current app-model does not provide a mechanism for this as far as I know. 



> You can't really remove the command registry from app-model. App-model at it's heart is just 3 registries (menu, command and keybinding registries)

Bad wording. I mean CommandID.



> There is an `enablement` field which lets you specify condition(s) for when the `Action` is active. However, maybe there is another use case example that is not satisfied by `Action`s?

It still works good only if you use napari as an application, not from a notebook or script. No one will restart a heavy notebook with calculations to modify a shortcut. 



> More details needed. An action that gets added to the menu (I don't think this is possible now even)? Would this be akin to NAP-6, but you want the user to be able to do this via script (on top of via a plugin)

This is not about adding actions to the menu, but general adding actions to napari (to be triggered by shortcut, for example, or by some custom controller)

**Comment by lucyleeow** [comment]
Okay that clarifies a lot thanks

**Comment by lucyleeow** [comment]
Let me summarise and expand.

### Too many files/Remove `_commands.py`

We need to edit too many files when amending/adding an `Action`. We could remove the `_commands.py` file, and instead just directly add a string command id and title when defining `Actions`. `app-model` will tell us if there is a duplicate command id.

Note I see [here](https://github.com/napari/napari/pull/6204#pullrequestreview-1640943601) you suggest an alternate syntax for `Action`s that uses command ID's. Maybe this concern could be handled later, once the app-model migration is in a more settled state?

I'm okay with this overall, I like being able to see all the IDs and titles in one place, but it's nice having one less file. @jni can you weigh in?

### Keybindings not editable

> Shortcuts are defined currently in Action, but it prevents them from being editable. 

@Czaki can you expand on exactly what you mean by this?

Discussion with @DragaDoncila about keybinding implementation (this should probably be moved somewhere else, just not sure where yet):

Shortcuts will be editable at runtime and the plan is for them to be able to be edited by the user. Currently you can update the app-model keybinding registry at runtime BUT this will not over-ride an existing keybinding (will require a PR in app-model to update this).

There are 3 levels of keybinding setting, from lowest to highest priority:

1. Native napari app
2. Plugins will be able to set keybindings
   * Implemented in https://github.com/napari/napari/pull/6204
3. User can set keybindings

The app-model `app` keybinding registry should store 'source' information (similar to VS code). The app will use the keybinding from the highest priority source. We will know when a keybinding comes from native napari or plugin (via manifest). Any setting of keybinding at runtime will be user set. The user should be able to do this via GUI or a new API function in headless mode (as we do not want to user to have to set 'source'). This addition will require a PR in app-model. When creating the app/keybinding registry, we should be able to set the possible 'source's and what priority they will be.

Note that keybinding rules also has a weight parameter (you could have >1 keybinding from the same source with different weights) but we don't think we want to expose this and would go with YAGNI.

### Dependency injection

> Currently, the menu are built using QModelMenu class from app-model. It uses injection every time use click, even if it could be resolved on constructor call. So bugs like https://github.com/napari/napari/issues/6466 create much more problem for end user.

I think we often do want to use dependency injection in actions, e.g., we want the current viewer not a viewer present when the action was registered. For cases where we do not want dependency injection, it is reasonably easy to pass the dependency to the callback (e.g., via a partial).

With regard to the bug in https://github.com/napari/napari/issues/6466, the problem would still have manifested but somewhere different? And in that case I think we would want to close the current viewer, not the viewer present at action registration time?

### Decorator

> We could adopt the old approach of using a decorator, which will add __action__ attribute to the function. Then, based on a list of all registered actions, the list of default shortcuts will be created.

AFAICT this decorator could be added to callbacks, turning them into an `Action` that is automatically (?) registered? (maybe it gets added to a list of `Action`s that are registered later?)

I think I am -0.5 on decorator. The locality problem (as mentioned by @DragaDoncila, I would prefer not to have to search and regex to find all actions in the e.g., file menu) and because I think it adds unnecessary complexity to the code.

**Comment by jni** [comment]
> We need to edit too many files when amending/adding an Action. We could remove the _commands.py file, and instead just directly add a string command id and title when defining Actions. app-model will tell us if there is a duplicate command id.

Yes, I think this is the right way forward. Strings are easier to read and easier for new contributors to edit. Re:

> I like being able to see all the IDs and titles in one place,

it would be trivial to generate such a file in a script, which could go in the tools dir.

> Note I see [here](https://github.com/napari/napari/pull/6204#pullrequestreview-1640943601) you suggest an alternate syntax for Actions that uses command ID's. Maybe this concern could be handled later, once the app-model migration is in a more settled state?

imho, yes. I'm very pro-strings and removing the CommandId class, and I think it will go a very long way towards simplifying contributing actions.

**Comment by tlambert03** [comment]
> Currently, the menu are built using QModelMenu class from app-model. It uses injection every time use click, even if it could be resolved on constructor call. So bugs like https://github.com/napari/napari/issues/6466 create much more problem for end user.

@Czaki, since you mention this issue a lot, i'll just point out that your provider for `Viewer` (`get_current_viewer`) sometimes returns `None`.  So, if someone (including napari) registers a callback function that requires a `Viewer` object to be injected, it's either up to them to provide a callback that can sometimes accept`None`, or it's up to you (napari) to provide a provider that always returns a `Viewer`.  

app-model is simply reporting the problem that "this function says it needs a Viewer, and the provider I've been given by napari hasn't returned one).  I don't see this so much as an issue about injection, it's about the design of the function that is being injected into.  Even if you *don't* use injection there, you'll still have to decide what to do when calling a user-provided function that has requested a viewer when you don't have one.

**Comment by lucyleeow** [comment]
I propose that we remove the `napari/_app_model/_submenus.py` file and define submenus in the same file as their actions. E.g., layer list submenus would be defined with the layerlist actions. Submenus will still be registered in the same place (on `NapariApplication` init)

I think it is not immediately clear where the submenu comes from, that it needs to be defined in a separate file and the enum adds confusion.

See zulip thread: https://napari.zulipchat.com/#narrow/stream/212875-general/topic/adding.20action.20to.20menubar/near/422508015 for confusion on 

cc @jni @Czaki @DragaDoncila

**Comment by lucyleeow** [comment]
Actually @psobolewskiPhD WDYT since you've had experience dealing with app-model. This would remove one file and thus another place you have to modify.

@Czaki WDYT ? :grimacing:

**Comment by DragaDoncila** [comment]
> I propose that we remove the napari/_app_model/_submenus.py file and define submenus in the same file as their actions. E.g., layer list submenus would be defined with the layerlist actions. Submenus will still be registered in the same place (on NapariApplication init)

I think this sounds perfectly sane. Would like to see the PR but I can't imagine it being more confusing than having to find some random file

**Comment by psobolewskiPhD** [comment]
> Actually @psobolewskiPhD WDYT since you've had experience dealing with app-model. This would remove one file and thus another place you have to modify.

I'm in favor of having submenus not together in a  _submenus file with their menu actions. I think it's much more logical and easier to understand the structure of things when reading code, let alone contributing.

**Comment by lucyleeow** [comment]
Thanks! I'll put in a PR after #6777 has been merged to avoid conflicts.

**Comment by lucyleeow** [comment]
@Czaki how do you feel about closing this for now. I have:

* Removed the command ID enums, so we no longer have a file listing all the commands
* Removed the separate submenu definition file, submenus are now defined with their actions

When adding a action we need to change:

* the file where the action is defined
* *if* the callback is defined elsewhere, we would need to change a 2nd file. I think more than half the time the callback is defined in the same file as the action, so often there is only one file to amend.

I still like that actions are defined in one file together. I also feel that a decorator adds reduces readability and for someone new to the codebase, it is less obvious where actions are defined. BUT if we later feel like we want to use a decorator or change where action and it's associated items are defined, we could open a new issue?

**Comment by lucyleeow** [comment]
I'm going to close this at the end of the week unless there are any objections. Thank you!
